sui/src/rpc_gateway.rs

```diff
@@ -0,0 +1,407 @@
+
+use crate::config::PersistedConfig;
+use crate::gateway::GatewayConfig;
+use crate::rest_gateway::responses::{NamedObjectRef, ObjectResponse};
+
+#[rpc(server, client, namespace = "sui")]
+pub trait RpcGateway {
+    #[method(name = "getObjectInfo")]
+    async fn get_object_info(&self, object_id: ObjectID) -> RpcResult<ObjectRead>;
+
+    #[method(name = "transferCoin")]
+    async fn transfer_coin(
+        &self,
+        signer: SuiAddress,
+        object_id: ObjectID,
+        gas_payment: ObjectID,
+        gas_budget: u64,
+        recipient: SuiAddress,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "moveCall")]
+    async fn move_call(
+        &self,
+        signer: SuiAddress,
+        package_object_id: ObjectID,
+        module: Identifier,
+        function: Identifier,
+        type_arguments: Vec<TypeTag>,
+        pure_arguments: Vec<Base64EncodedBytes>,
+        gas_object_id: ObjectID,
+        gas_budget: u64,
+        object_arguments: Vec<ObjectID>,
+        shared_object_arguments: Vec<ObjectID>,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "publish")]
+    async fn publish(
+        &self,
+        sender: SuiAddress,
+        compiled_modules: Vec<Base64EncodedBytes>,
+        gas_object_id: ObjectID,
+        gas_budget: u64,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "splitCoin")]
+    async fn split_coin(
+        &self,
+        signer: SuiAddress,
+        coin_object_id: ObjectID,
+        split_amounts: Vec<u64>,
+        gas_payment: ObjectID,
+        gas_budget: u64,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "mergeCoins")]
+    async fn merge_coin(
+        &self,
+        signer: SuiAddress,
+        primary_coin: ObjectID,
+        coin_to_merge: ObjectID,
+        gas_payment: ObjectID,
+        gas_budget: u64,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "executeTransaction")]
+    async fn execute_transaction(
+        &self,
+        signed_transaction: SignedTransaction,
+    ) -> RpcResult<TransactionResponse>;
+
+    #[method(name = "syncAccountState")]
+    async fn sync_account_state(&self, address: SuiAddress) -> RpcResult<()>;
+
+    #[method(name = "getOwnedObjects")]
+    async fn get_owned_objects(&self, owner: SuiAddress) -> RpcResult<ObjectResponse>;
+
+    #[method(name = "getTotalTransactionNumber")]
+    async fn get_total_transaction_number(&self) -> RpcResult<u64>;
+
+    #[method(name = "getTransactionsInRange")]
+    async fn get_transactions_in_range(
+        &self,
+        start: GatewayTxSeqNumber,
+        end: GatewayTxSeqNumber,
+    ) -> RpcResult<Vec<(GatewayTxSeqNumber, TransactionDigest)>>;
+
+    #[method(name = "getRecentTransactions")]
+    async fn get_recent_transactions(
+        &self,
+        count: u64,
+    ) -> RpcResult<Vec<(GatewayTxSeqNumber, TransactionDigest)>>;
+}
+
+pub struct RpcGatewayImpl {
+    gateway: Arc<Mutex<GatewayClient>>,
```

Comment:
One thing that isn't very clear to me is what purpose the GatewayClient serves and why it requires being wrapped in a mutex. My main concern with this design paradigm is that you are unintentionally limiting the number of requests that you can process concurrently to 1 since most of the APIs below acquire the lock which is then not released until the end of the RPC method call (which can spawn multiple await points).

Reply:
In general, the `gateway` requires a `.lock()` even for read-only calls, which doesn't seem like it will scale even to the limited degree we need gateways to scale. +1 to this concern

Reply:
Thanks for the review!! The Mutex wrap was because of the `&mut` requirement from some of the GatewayAPI functions, but we don't need them anymore after the datastore and aggregator refactoring; I have removed them and removed the ArcMutex wrap.